### PR TITLE
feat: add CU worker binary

### PR DESCRIPTION
**Pull request by monoid**

### CODE REVIEW

File: cu_main/docker/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        /tmp/* \
+        /var/lib/apt/lists/* \
+        /var/tmp/*
+
+FROM user-builder AS ccp_main
+COPY --from=ccp-builder /tmp/ccp_cu /usr/local/bin/ccp_cu
+COPY cu_main/docker/Config.default.toml /Config.toml
+COPY cu_main/docker/entrypoint.sh /entrypoint.sh
+USER ccp_user
+ENTRYPOINT ["/usr/local/bin/ccp_cu"]
```
Comment (monoid) [inline cu_main/docker/Dockerfile]: It should be `/entrypoint.sh`.

File: Cargo.toml [hunk 2]
```diff
@@ -18,6 +18,7 @@ members = [
     "crates/test-utils",
     "crates/utils",
     "main",
+    "cu_main", "crates/cu-rpc-client",
```
Comment (monoid) [inline Cargo.toml]: Fix formatting.

File: crates/config/src/cu_config.rs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+use crate::defaults::default_utility_queue_size;
+use crate::unresolved_config::*;
+use crate::*;
+
+/// Subset of main config for ccp-cu-main worker.
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct CUWorkerConfig {
+    pub optimizations: Optimizations,
+    // TODO remove
```
Comment (monoid) [inline crates/config/src/cu_config.rs]: Remove.

File: crates/cu-rpc-client/proto/collector.proto [hunk 4]
```diff
@@ -0,0 +1,27 @@
+syntax = "proto3";
+
+package cu_rpc_client;
+
+message EpochParameters {
+    string difficulty = 1;
```
Comment (gurinderu) [inline crates/cu-rpc-client/proto/collector.proto:6]: Lets use bytes everywhere for the byte arrays.

Comment (monoid) [reply crates/cu-rpc-client/proto/collector.proto:6]: I think protobuf suxx in Rust projects and should be removed. It gives you foreign, unergonomic, un-Rusty types.

File: cu_worker/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+// same as ccp_main crate
+const CCP_LOG_ENV_VAR: &str = "CCP_LOG";
+
+#[derive(Envconfig)]
+struct TaskParameters {
+    global_nonce: GlobalNonce,
+    difficulty: Difficulty,
+    cu_id: CUID,
+    upstream_url: String,
```
Comment (gurinderu) [inline cu_worker/src/main.rs]: Url ?

**Review by gurinderu — APPROVED** [review]

File: crates/cu-rpc-client/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+                    Ok::<_, std::io::Error>(TokioIo::new(UnixStream::connect(path).await?))
+                }
+            }))
+            .await?;
+        let inner = CollectorClient::new(channel);
+        Ok(Self { inner })
+    }
+
+    pub async fn auto(endpoint: String) -> Result<Self, tonic::transport::Error> {
+        if endpoint.starts_with("http://") || endpoint.starts_with("https://") {
```
Comment (gurinderu) [inline crates/cu-rpc-client/src/lib.rs]: grpc url typically start with `grpc://`

File: crates/cu-rpc-client/proto/collector.proto (hunk 4 above)
Comment (monoid) [reply crates/cu-rpc-client/proto/collector.proto:6]: Using bytes would involve writing more code without real benefits.

File: cu_worker/src/main.rs (hunk 5 above)
Comment (monoid) [reply cu_worker/src/main.rs]: Renamed to upstream.

File: .github/workflows/artifact.yml [hunk 7]
```diff
@@ -203,4 +203,4 @@ jobs:
           vault read kv/basicauth/dev
           docker login docker.fluence.dev
           ```
-          SNAPSHOT
+          SNAPSHOT
```
Comment (gurinderu) [inline .github/workflows/artifact.yml]: \n at eof

**Review by gurinderu — APPROVED** [review]

**Pull request merged**